server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java

```diff
@@ -1061,4 +960,104 @@ public void putMemTableBackAndClose() throws TsFileProcessorException {
   public ReadWriteLock getFlushQueryLock() {
     return flushQueryLock;
   }
+
+  private File createNewTmpFile() {
+    File parent = writer.getFile().getParentFile();
+    return FSFactoryProducer.getFSFactory().getFile(parent,
+        writer.getFile().getName() + IoTDBConstant.FILE_NAME_SEPARATOR + System
+            .currentTimeMillis()
+            + VM_SUFFIX + IoTDBConstant.PATH_SEPARATOR
+            + PATH_UPGRADE);
+  }
+
+  private void flushAllVmToTsFile(List<RestorableTsFileIOWriter> currMergeVmWriters,
+      List<TsFileResource> currMergeVmFiles) throws IOException {
+    VmMergeTask vmMergeTask = new VmMergeTask(writer, currMergeVmWriters,
+        storageGroupName,
+        new VmLogger(tsFileResource.getFile().getParent(),
+            tsFileResource.getFile().getName()),
+        new HashSet<>(), sequence);
+    vmMergeTask.fullMerge();
+    for (TsFileResource vmTsFileResource : currMergeVmFiles) {
+      deleteVmFile(vmTsFileResource);
+    }
+    vmWriters.removeAll(currMergeVmWriters);
+    vmTsFileResources.removeAll(currMergeVmFiles);
+    File logFile = FSFactoryProducer.getFSFactory()
+        .getFile(tsFileResource.getFile().getParent(),
+            tsFileResource.getFile().getName() + VM_LOG_NAME);
+    logFile.delete();
+  }
+
+  class VmMergeLoop implements Callable<Void> {
+
+    public void merge() {
+      try {
+        if (shouldClose && flushingMemTables.isEmpty()) {
+          // merge vm to tsfile
+          logger.info("[Flush] merge all {} vms to TsFile", vmTsFileResources.size() + 1);
+          flushAllVmToTsFile(vmWriters, vmTsFileResources);
+          return;
+        }
+
+        long vmPointNum = 0;
+        // all flush to target file
+        int mergeFileNum = vmWriters.size() - 1;
+        List<RestorableTsFileIOWriter> flushVmWriters = vmWriters.subList(0, mergeFileNum);
+        List<TsFileResource> flushVmFiles = vmTsFileResources.subList(0, mergeFileNum);
+        Map<Path, MeasurementSchema> pathMeasurementSchemaMap = new HashMap<>();
+        for (RestorableTsFileIOWriter vmWriter : flushVmWriters) {
+          Map<String, Map<String, List<ChunkMetadata>>> schemaMap = vmWriter
+              .getMetadatasForQuery();
+          for (String device : schemaMap.keySet()) {
+            for (String measurement : schemaMap.get(device).keySet()) {
+              List<ChunkMetadata> chunkMetadataList = schemaMap.get(device).get(measurement);
+              for (ChunkMetadata chunkMetadata : chunkMetadataList) {
+                vmPointNum += chunkMetadata.getNumOfPoints();
+              }
+              pathMeasurementSchemaMap.putIfAbsent(new Path(device, measurement),
```

Comment:
in this way, you will construct many Path and MeasurementSchema, use computeIfAbsent and Map<String, Map<String, MeasurementSchema>>